File: src/test/java/de/uni_mannheim/desq/mining/Icdm16TraditionalMiningTest.java
package de.uni_mannheim.desq.mining;

import de.uni_mannheim.desq.dictionary.Dictionary;
import de.uni_mannheim.desq.io.DelSequenceReader;
import de.uni_mannheim.desq.io.SequenceReader;
import de.uni_mannheim.desq.util.DesqProperties;
import org.apache.commons.lang3.tuple.Pair;
import org.junit.runners.Parameterized;

import java.io.IOException;
import java.net.URL;
import java.util.*;

/**
 * Created by rgemulla on 18.07.2016.
 */
public class Icdm16TraditionalMiningTest extends TraditionalMiningTest {
    public static Collection<Object[]> baseData() {
        List<Object[]> parameters = new ArrayList<>();
        for (Long sigma : new Long[] {1L,3L,5L,7L})
            for (Integer gamma : new Integer[] {0,1,2})
                for (Integer lambda : new Integer[] {1,3,5,7})
                    for (Boolean generalize : new Boolean[]{ false, true}) {
                        parameters.add(new Object[] { sigma, gamma, lambda, generalize });
                    }
        return parameters;
    }

    @Parameterized.Parameters(name = "Icdm16TraditionalMiningTest-{4}-{0}-{1}-{2}-{3}-")
    public static Collection<Object[]> data() {
        List<Object[]> parameters = new ArrayList<>();
        for (Object[] par : baseData()) {
            for (Pair<String, DesqProperties> miner : MinerConfigurations.all((Long)par[0], (Integer)par[1],
                    (Integer)par[2], (Boolean)par[3])) {
                parameters.add(new Object[]{par[0], par[1], par[2], par[3], miner.getLeft(), miner.getRight()});
            }
        }
        Collections.sort(parameters, (p1,p2) -> ((String)p1[4]).compareTo((String)p2[4]));
        return parameters;
    }

    public Icdm16TraditionalMiningTest(long sigma, int gamma, int lambda, boolean generalize,
                                       String minerName, DesqProperties conf) {
        super(sigma, gamma, lambda, generalize, minerName, conf);
    }

    @Override
    public Dictionary getDictionary() throws IOException {
        URL dictFile = TraditionalMiningTest.class.getResource("/icdm16-example/dict.json");
        return Dictionary.loadFrom(dictFile);
    }

    @Override
    public SequenceReader getSequenceReader() throws IOException {
        URL dataFile = TraditionalMiningTest.class.getResource("/icdm16-example/data.del");
        return new DelSequenceReader(dataFile.openStream(), false);
    }

    @Override
    public boolean computeStatisticsAndFids() {
        return true;
    }

    @Override
    public String getGoldFileBaseName() {
        return "icdm16/icdm16-traditional-patterns-ids";
    }

    @Override
    public String getTestDirectoryName() {
        return getClass().getSimpleName();
    }
}

File: src/test/scala/de/uni_mannheim/desq/mining/spark/Icdm16TraditionalMiningTest.scala
package de.uni_mannheim.desq.mining.spark

import java.util
import java.util.{Collections, Comparator}

import de.uni_mannheim.desq.dictionary.Dictionary
import de.uni_mannheim.desq.io.DelSequenceReader
import de.uni_mannheim.desq.mining.WeightedSequence
import de.uni_mannheim.desq.util.DesqProperties
import org.apache.spark.SparkContext
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

import scala.collection.JavaConversions._
import scala.io.Source

/**
 * Created by rgemulla on 18.07.2016.
 */
@RunWith(classOf[Parameterized])
class Icdm16TraditionalMiningTest(sigma: Long, gamma: Int, lambda: Int, generalize: Boolean,
                                  minerName: String, conf: DesqProperties)
  extends TraditionalMiningTest(sigma, gamma, lambda, generalize, minerName, conf) {

    override def goldFileBaseName = "icdm16/icdm16-traditional-patterns-ids"

    override def testDirectoryName = getClass.getSimpleName

    /** The data */
    override def getDesqDataset()(implicit sc: SparkContext): DesqDataset = Icdm16TraditionalMiningTest.getDesqDataset()
    override def getGenericDesqDataset()(implicit sc: SparkContext): GenericDesqDataset[(Array[String], Long)] = Icdm16TraditionalMiningTest.getGenericDesqDataset()
}

object Icdm16TraditionalMiningTest {
    var desqDataset: DesqDataset = _
    var genericDesqDataset: GenericDesqDataset[(Array[String], Long)] = _

    @Parameterized.Parameters(name = "Icdm16TraditionalMiningTest-{4}-{0}-{1}-{2}-{3}-")
    def data(): util.Collection[Array[Object]] = {
      val parameters = new util.ArrayList[Array[Object]]()
      val baseData = de.uni_mannheim.desq.mining.Icdm16TraditionalMiningTest.baseData()
      for (par <- collectionAsScalaIterable(baseData)) {
        for (miner <- MinerConfigurations.all(par(0).asInstanceOf[java.lang.Long],
          par(1).asInstanceOf[java.lang.Integer], par(2).asInstanceOf[java.lang.Integer],
          par(3).asInstanceOf[java.lang.Boolean])) {
            parameters.add(Array[Object](par(0), par(1), par(2), par(3), miner._1, miner._2))
        }
      }

      Collections.sort(parameters, new Comparator[Array[Object]] {
        override def compare(o1: Array[Object], o2: Array[Object]): Int = {
          o1(4).asInstanceOf[String].compareTo(o2(4).asInstanceOf[String])
        }
      })
      parameters
    }

    /** The data */
    def getDesqDataset()(implicit sc: SparkContext): DesqDataset = {
      if (desqDataset == null) {
        val dictFile = this.getClass.getResource("/icdm16-example/dict.json")
        val dataFile = this.getClass.getResource("/icdm16-example/data.del")

        // load the dictionary & update hierarchy
        val dict = Dictionary.loadFrom(dictFile)
        val delFile = sc.parallelize(Source.fromURL(dataFile).getLines.toSeq)
        desqDataset = DesqDataset.loadFromDelFile(delFile, dict, usesFids = false).recomputeDictionary()
        desqDataset.sequences.cache()
      }
      desqDataset
    }

    /** The data */
    def getGenericDesqDataset()(implicit sc: SparkContext): GenericDesqDataset[(Array[String], Long)] = {
      if (genericDesqDataset == null) {
        val desqDataset = getDesqDataset()
        val sequences = desqDataset.sequences.collect().map(ws => {
          (desqDataset.descriptor.getSids(ws), desqDataset.descriptor.getWeight(ws))
        })

        val descriptor = new StringArrayAndLongDescriptor()
        descriptor.setDictionary(desqDataset.descriptor.getDictionary)

        genericDesqDataset = new GenericDesqDataset[(Array[String], Long)](sc.parallelize(sequences), descriptor)
        genericDesqDataset.sequences.cache()
      }
      genericDesqDataset
    }

}